### ISSUE TITLE
代码问题

### ISSUE THREAD

**Issue by ZMZJ13** [MicroResearchLab/AMP-potency-prediction-EvoGradient#4]
作者您好，我注意到在开源代码库中，仅提供了 CNN、LSTM、Attention 和 Transformer 模型的训练权重，但缺少完整的训练代码。因此，我非常希望能够获得这些模型的训练代码，以便于更好地理解和使用这些模型。

**Comment by SYYatSEU** [comment]
您好，目前我们已经提供的源码已经足够您进行完整的复现。
若您需要进行新模型的训练，相关的所有必要内容已十分充足，可以自行构建训练。
- CNN、LSTM、Transformer 模型定义在`otherModels5_2Output.py`和`otherModelsRegression.py`中，Attention模型定义则在`AMP_classification.py#L185`中；
- 两个任务的数据集定义分别在`AMP_classification.py#L114``AMP_regression.py#L99`；
- 训练的loss函数的参考实现：
  ```python
  class LossWithPenalty(nn.Module):
      def __init__(self, lambda_):
          super(LossWithPenalty, self).__init__()
          self.lambda_ = lambda_
  
      def forward(self, logits, target):
          # Assuming target is a batch of binary labels (0 or 1)
          batch_size = target.size(0)
          target = target.long()
          # Create the encoded target for the batch
          encoded_target = torch.zeros((batch_size, 2), device=logits.device)
          encoded_target.scatter_(1, target.unsqueeze(1), 1)
          # Calculate original loss
          ori_loss = nn.CrossEntropyLoss()(logits, target)
          # Calculate penalty
          score = F.softmax(logits, dim=-1)[:, 1]  # Get the probability of the positive class
          penalty = torch.sum(torch.square(F.relu(score - target.float())))
          # Calculate total loss
          total_loss = ori_loss + self.lambda_ * penalty
          return total_loss
  ```
- 训练超参数num of epochs、batch size、learning rate、optimizer、λ均可在论文中找到我们的设定，您也可以根据您的数据集情况自行设定。
